File: modules/swagger-core/src/test/scala/1_3/converter/ModelPropertyTest.scala
package converter

import java.util.Date

import com.fasterxml.jackson.module.scala.DefaultScalaModule
import io.swagger.converter.ModelConverters
import io.swagger.models.properties._
import io.swagger.util.Json
import matchers.SerializationMatchers._
import models._
import org.junit.runner.RunWith
import org.scalatest.{FlatSpec, Matchers}
import org.scalatest.junit.JUnitRunner

@RunWith(classOf[JUnitRunner])
class ModelPropertyTest extends FlatSpec with Matchers {
  Json.mapper().registerModule(DefaultScalaModule)

  it should "extract properties" in {
    val models = ModelConverters.getInstance().readAll(classOf[Family])
    models.size should be(3)
    val person = models.get("Person")
    val employer = person.getProperties().get("employer")

    employer.isInstanceOf[ArrayProperty] should be(true)
    val employerProperty = employer.asInstanceOf[ArrayProperty]

    val items = employerProperty.getItems
    items.isInstanceOf[RefProperty] should be(true)
    val itemsProperty = items.asInstanceOf[RefProperty]
    itemsProperty.getSimpleRef() should equal("Employer")

    val awards = person.getProperties().get("awards")
    awards.isInstanceOf[ArrayProperty] should be(true)
    val awardsProperty = awards.asInstanceOf[ArrayProperty]
    awardsProperty.getItems().isInstanceOf[StringProperty] should be(true)
  }

  it should "extract a primitive array" in {
    val models = ModelConverters.getInstance().readAll(classOf[ModelWithPrimitiveArray])

    models.size should be(1)
    val model = models.get("ModelWithPrimitiveArray")
    val longArray = model.getProperties().get("longArray").asInstanceOf[ArrayProperty]
    val longArrayItems = longArray.getItems()
    longArrayItems.isInstanceOf[LongProperty] should be(true)

    val intArray = model.getProperties().get("intArray").asInstanceOf[ArrayProperty]
    val awardItems = intArray.getItems()
    awardItems.isInstanceOf[IntegerProperty] should be(true)
  }

  it should "read a model property" in {
    val models = ModelConverters.getInstance().readAll(classOf[IsModelTest])
    val model = models.get("IsModelTest")
    model should not be (null)
  }

  it should "read a scala object" in {
    val models = ModelConverters.getInstance().readAll(classOf[Pet])
    models.size should be(1)
    val model = models.get("Pet")
    model.getProperties().get("name") should not be (null)
  }
}

@RunWith(classOf[JUnitRunner])
class ModelPropertyOverrideTest extends FlatSpec with Matchers {
  it should "read a model with property dataTypes configured #679" in {
    val models = ModelConverters.getInstance().readAll(classOf[ModelWithModelPropertyOverrides])
    models should serializeToJson(
      """{
  "Children" : {
    "type": "object",
    "properties" : {
      "name" : {
        "type" : "string"
      }
    }
  },
  "ModelWithModelPropertyOverrides" : {
    "type": "object",
    "properties" : {
      "children" : {
        "type" : "array",
        "items" : {
          "$ref" : "#/definitions/Children"
        }
      }
    }
  }
}""")
  }
}

case class Family(membersSince: Date, members: List[Person])

class Pet {
  var name: String = _
  var age: Int = 0
  var birthday: java.util.Date = _
}

case class Person(
                   firstname: String,
                   lastname: String,
                   middlename: Option[String],
                   age: Int,
                   birthday: Date,
                   employer: List[Employer],
                   awards: List[String])

case class Employer(
                     name: String,
                     size: Int)

case class IsModelTest(
                        is_happy: Boolean,
                        name: String)

File: modules/swagger-core/src/test/java/io/swagger/ModelPropertyTest.java
package io.swagger;

import io.swagger.converter.ModelConverters;
import io.swagger.matchers.SerializationMatchers;
import io.swagger.models.Model;
import io.swagger.models.Model1979;
import io.swagger.models.ModelImpl;
import io.swagger.models.ModelWithBooleanProperty;
import io.swagger.models.ModelWithModelPropertyOverrides;
import io.swagger.models.ModelWithPrimitiveArray;
import io.swagger.models.ReadOnlyFields;
import io.swagger.models.properties.ArrayProperty;
import io.swagger.models.properties.BooleanProperty;
import io.swagger.models.properties.IntegerProperty;
import io.swagger.models.properties.LongProperty;
import io.swagger.models.properties.Property;
import io.swagger.models.properties.RefProperty;
import io.swagger.models.properties.StringProperty;
import org.testng.annotations.Test;

import java.util.Date;
import java.util.List;
import java.util.Map;

import static org.testng.Assert.assertEquals;
import static org.testng.Assert.assertNotNull;
import static org.testng.Assert.assertTrue;

public class ModelPropertyTest {
    @Test
    public void extractProperties() {
        final Map<String, Model> models = ModelConverters.getInstance().readAll(Family.class);
        assertEquals(models.size(), 3);

        final Model person = models.get("Person");
        final Property employer = person.getProperties().get("employer");

        assertTrue(employer instanceof ArrayProperty);
        final ArrayProperty employerProperty = (ArrayProperty) employer;

        final Property items = employerProperty.getItems();
        assertTrue(items instanceof RefProperty);
        assertEquals(((RefProperty) items).getSimpleRef(), "Employer");

        final Property awards = person.getProperties().get("awards");
        assertTrue(awards instanceof ArrayProperty);
        assertTrue(((ArrayProperty) awards).getItems() instanceof StringProperty);
    }

    @Test
    public void extractPrimitiveArray() {
        final Map<String, Model> models = ModelConverters.getInstance().readAll(ModelWithPrimitiveArray.class);
        assertEquals(models.size(), 1);

        final Model model = models.get("ModelWithPrimitiveArray");
        final ArrayProperty longArray = (ArrayProperty) model.getProperties().get("longArray");
        final Property longArrayItems = longArray.getItems();
        assertTrue(longArrayItems instanceof LongProperty);

        final ArrayProperty intArray = (ArrayProperty) model.getProperties().get("intArray");
        assertTrue(intArray.getItems() instanceof IntegerProperty);
    }

    @Test
    public void readModelProperty() {
        final Map<String, Model> models = ModelConverters.getInstance().readAll(IsModelTest.class);
        final Model model = models.get("IsModelTest");
        assertNotNull(model);
    }

    @Test(description = "it should read a model with property dataTypes configured #679")
    public void readDataTypesProperty() {
        final Map<String, Model> models = ModelConverters.getInstance().readAll(ModelWithModelPropertyOverrides.class);
        final String json = "{" +
                "   \"Children\":{" +
                "      \"type\":\"object\"," +
                "      \"properties\":{" +
                "         \"name\":{" +
                "            \"type\":\"string\"" +
                "         }" +
                "      }" +
                "   }," +
                "   \"ModelWithModelPropertyOverrides\":{" +
                "      \"type\":\"object\"," +
                "      \"properties\":{" +
                "         \"children\":{" +
                "            \"type\":\"array\"," +
                "            \"items\":{" +
                "               \"$ref\":\"#/definitions/Children\"" +
                "            }" +
                "         }" +
                "      }" +
                "   }" +
                "}";
        SerializationMatchers.assertEqualsToJson(models, json);
    }

    @Test
    public void testReadOnlyProperty() {
        final Map<String, Model> models = ModelConverters.getInstance().readAll(ReadOnlyFields.class);
        ModelImpl model = (ModelImpl) models.get("ReadOnlyFields");
        assertTrue(model.getProperties().get("id").getReadOnly());
    }

    @Test
    public void modelAllowEmptyTest() {
        final Map<String, Model> models = ModelConverters.getInstance().readAll(Model1979.class);
        ModelImpl model = (ModelImpl) models.get("Model1979");
        assertTrue(model.getProperties().get("id").getAllowEmptyValue());
    }


    @Test
    public void testIssue1743() {
        final Map<String, Model> models = ModelConverters.getInstance().readAll(ModelWithBooleanProperty.class);
        final Model model = models.get("ModelWithBooleanProperty");
        assertNotNull(model);

        BooleanProperty bp = (BooleanProperty) model.getProperties().get("isGreat");
        assertTrue(bp.getEnum().size() == 1);
        assertEquals(bp.getEnum().get(0), Boolean.TRUE);
    }

    class Family {
        public Date membersSince;
        public List<Person> members;
    }

    class Person {
        public String firstname;
        public String lastname;
        public int age;
        public Date birthday;
        public List<Employer> employer;
        public List<String> awards;
    }

    class Employer {
        public String name;
        public int size;
    }

    class IsModelTest {
        public Boolean is_happy;
        public String name;
    }
}
